src/checkout/views/Shipping/View.tsx

```diff
@@ -0,0 +1,215 @@
+import { History } from "history";
+import * as React from "react";
+import { generatePath, RouteComponentProps } from "react-router";
+
+import {
+  FormAddressType,
+  OverlayContext,
+  OverlayContextInterface,
+  OverlayTheme,
+  OverlayType,
+  ShippingAddressForm
+} from "../../../components";
+import { FormError } from "../../../components/Form";
+import { ShopContext } from "../../../components/ShopProvider/context";
+import { getShop_shop } from "../../../components/ShopProvider/types/getShop";
+import { maybe } from "../../../core/utils";
+
+import {
+  CartContext,
+  CartLineInterface
+} from "../../../components/CartProvider/context";
+import { Steps } from "../../components";
+import {
+  CheckoutContext,
+  CheckoutContextInterface,
+  CheckoutStep
+} from "../../context";
+import { TypedCreateCheckoutMutation } from "../../queries";
+import { shippingOptionsUrl } from "../../routes";
+import { Checkout } from "../../types/Checkout";
+import {
+  createCheckout,
+  createCheckout_checkoutCreate
+} from "../../types/createCheckout";
+import { TypedUpdateCheckoutShippingAddressMutation } from "./queries";
+import ShippingUnavailableModal from "./ShippingUnavailableModal";
+import {
+  updateCheckoutShippingAddress,
+  updateCheckoutShippingAddress_checkoutShippingAddressUpdate
+} from "./types/updateCheckoutShippingAddress";
+
+const proceedToShippingOptions = (
+  update: (checkoutData: CheckoutContextInterface) => void,
+  history: History,
+  overlay: OverlayContextInterface,
+  token?: string
+) => (
+  data:
+    | createCheckout_checkoutCreate
+    | updateCheckoutShippingAddress_checkoutShippingAddressUpdate
+) => {
+  const canProceed = !data.errors.length;
+
+  if (canProceed) {
+    const shippingUnavailable = !data.checkout.availableShippingMethods.length;
+
+    if (shippingUnavailable) {
+      overlay.show(OverlayType.modal, OverlayTheme.modal, {
+        content: <ShippingUnavailableModal hide={overlay.hide} />
+      });
+    } else {
+      update({
+        checkout: data.checkout,
+        step: CheckoutStep.ShippingOption
+      });
+      history.push(generatePath(shippingOptionsUrl, { token }));
+    }
+  }
+};
+
+const extractShippingData = (checkout: Checkout, shop: getShop_shop) => {
+  const hasShippingCountry = maybe(() => !!checkout.shippingAddress.country);
```

Comment:
I know it's a tiny thing, but `hasShippingCountry` can be `true`, `false` or `undefined` if you use `!!` before `maybe` expression rather than before `maybe` itself.

Reply:
`undefined` satisfies any type by default. `maybe` will fall back to `undefined` unless you provide the fallback value as the second parameter.

Reply:
Looking through the definition I thought so too. Just got confused that intelisense claimed otherwise.